### PR TITLE
feat: verify contracts on sourcify during deploy

### LINKED ISSUE #1983: Verify contracts in sourcify on deploy
**Is your feature request related to a problem? Please describe.**
Following our last regenesis on 14th November 2021 we verified contracts in both `optimistic-kovan` and `optimistic-mainnet` in `sourcify` and added support for optimism networks there:
https://github.com/ethereum/sourcify/pull/550
https://github.com/ethereum/sourcify/pull/581

This process was manual for the regenesis however for future deployments we should automate that as part of the deploy process. 

**Describe the solution you'd like**
Update the `DeployAndVerifyAndThen` script to add `sourcify` verification step.

**Additional context**
#1722 Adds source code verification on Etherscan during the deployment process. 
#1726 add better error handling into that process. 
#1534 Adds metadata settings to solc to enable contract verification on Sourcify

### PR DESCRIPTION
**Pull request by rajivpo**
***This has not yet been tested.***

**Description**
Sourcify verification of contracts during deployment. 

**Additional context**
This was surfaced as something that needed to be handled manually during the latest regeneses.

**Metadata**
- Fixes #1983

### CODE REVIEW

**Comment by rajivpo** [comment]
Not sure what the best way to test this would be. Open to any suggestions cc: @elenadimitrova

File: packages/contracts/src/deploy-utils.ts [hunk 1]
```diff
@@ -53,9 +53,18 @@ export const deployAndVerifyAndThen = async ({
...
+        console.log('Error when verifying bytecode on Etherscan:')
+        console.log(error)
+      }
+
+      try {
+        console.log('Verifying on Sourcify...')
+        await hre.run('sourcify')
+        console.log('Successfully verified on Sourcify')
+      } catch (error) {
+        console.log('Error when verifying bytecode on Etherscan:')
```
Comment (indeavr) [inline packages/contracts/src/deploy-utils.ts]: quick fix -> 'Error when verifying bytecode on `Sourcify`'

**Comment by elenadimitrova** [comment]
@rajivpo it will need to be a manual test against kovan where we run one of the deploy scripts which uses `deployAndVerifyAndThen`. Once you mark this ready-to-review I'll test.

**Comment by rajivpo** [comment]
@elenadimitrova I think this is ready to review if you have a bit of time 😄

**Review by Rose2161 — APPROVED** [review]

File: packages/contracts/src/deploy-utils.ts [hunk 2]
```diff
@@ -53,9 +53,18 @@ export const deployAndVerifyAndThen = async ({
...
+        console.log('Successfully verified on Etherscan')
       } catch (error) {
-        console.log('Error when verifying bytecode on etherscan:')
+        console.log('Error when verifying bytecode on Etherscan:')
+        console.log(error)
+      }
+
+      try {
+        console.log('Verifying on Sourcify...')
+        await hre.run('sourcify')
```
Comment (elenadimitrova) [inline packages/contracts/src/deploy-utils.ts:64]: This required address to verify and constructor arguments, same as etherscan.

```
await hre.run("sourcify", {
                address: contract,
                constructorArguments: [...]
            });
```

**Review by elenadimitrova — CHANGES_REQUESTED** [review]

File: packages/contracts/src/deploy-utils.ts (hunk 2 above)
Comment (rajivpo) [reply packages/contracts/src/deploy-utils.ts:64]: Interesting. I might be missing something, but it looks like the [`sourcify` task grabs the address and relevant metadata from `hre.deployments`](https://github.com/wighawag/hardhat-deploy/blob/2b1ade5b69b740b0d808a04370b31fa2e61d306f/src/sourcify.ts#L37-L115).

**Review by elenadimitrova — APPROVED** [review]

**Comment by elenadimitrova** [comment]
@rajivpo yes sorry was just waiting for a rebase. Thanks for adding this in!

**Pull request merged**